File: FullLessons/subject5/python/MazeSolver.py
#!/usr/bin/env python3
import os
import platform
import time


def load_maze(filename):
    """
    Loads a maze from a text file in the resources directory.

    Args:
        filename (str): The name of the maze file (without .txt).

    Returns:
        tuple: A tuple containing the maze (list of lists), start coordinates (x, y), and end coordinates (x, y).
    """
    try:
        with open(f"resources/mazes/{filename}.txt", "r") as f:
            rows, cols = map(int, f.readline().split())
            maze = [list(f.readline().strip()) for _ in range(rows)]
            start_x, start_y = None, None
            end_x, end_y = None, None
            for i in range(rows):
                for j in range(cols):
                    if maze[i][j] == "S":
                        start_x, start_y = j, i
                    elif maze[i][j] == "E":
                        end_x, end_y = j, i
            return maze, start_x, start_y, end_x, end_y
    except FileNotFoundError:
        print(f"Maze file not found: resources/{filename}.txt")
        exit(1)


def print_maze(maze):
    """Prints the maze to the console."""
    for row in maze:
        print("".join(row))


def clear_console():
    """Clears the console."""
    if platform.system() == "Windows":
        os.system("cls")
    else:
        os.system("clear")


# --- Main program ---

# Load the maze
filename = input("Enter maze filename (without .txt): ")
# filename = "Maze1"  # For testing with a specific maze file
maze, start_x, start_y, end_x, end_y = load_maze(filename)

# Initialize current position
current_x, current_y = start_x, start_y

# Clear the console and print the initial maze
clear_console()
print_maze(maze)

loop_iteration = 0

# Main loop
while current_x != end_x or current_y != end_y:
    loop_iteration += 1

    # Get user input for the move
    move = input("Enter move (w/a/s/d): ").lower()

    new_x, new_y = current_x, current_y

    # Update position based on the move
    if move == "w":
        new_y -= 1
    elif move == "a":
        new_x -= 3  # Adjusted for horizontal movement
    elif move == "s":
        new_y += 1
    elif move == "d":
        new_x += 3  # Adjusted for horizontal movement
    else:
        print("Invalid move.")
        continue

    try:
        # Move to the nearest point to the wall
        while maze[new_y][new_x] == "#":
            if move == "w":
                new_y += 1
            elif move == "a":
                new_x += 1
            elif move == "s":
                new_y -= 1
            elif move == "d":
                new_x -= 1
            else:
                raise ValueError("Error with movement")

        # Keep the position within the maze boundaries
        new_x = max(0, min(new_x, len(maze[0]) - 2))
        new_y = max(0, min(new_y, len(maze) - 2))

        # Update the maze and current position
        clear_console()
        maze[current_y][current_x] = " "
        current_x, current_y = new_x, new_y
        maze[current_y][current_x] = "t"

    except IndexError:
        # Handle the case where the new position is out of bounds
        new_x = max(0, min(new_x, len(maze[0]) - 2))
        new_y = max(0, min(new_y, len(maze) - 2))

    # Print the updated maze
    print_maze(maze)
    time.sleep(0.2)  # Add a small delay for better visualization

print(f"Congratulations! You solved the maze in {loop_iteration} loop iterations.")

File: FullLessons/subject5/src/main/java/com/hse/programming/MazeSolver.java
package com.hse.programming;

import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.rmi.UnexpectedException;
import java.util.Scanner;

public class MazeSolver {

  private static char[][] maze;
  private static int startX, startY;
  private static int endX, endY;
  private static int currentX, currentY; // Track current position

  public static void main(String[] args) throws UnexpectedException {
    Scanner scanner = new Scanner(System.in);
    System.out.print("Enter maze filename (without .txt): ");
    String filename = scanner.nextLine();

    loadMaze(filename);
    // Clear the console before printing the initial maze
    clearConsole();
    printMaze();

    int loopiteration = 0;

    currentX = startX;
    currentY = startY;

    while (currentX != endX || currentY != endY) {
      loopiteration++;

      // Ask user where to move
      System.out.println("Enter move (w/a/s/d): ");
      char move = ' ';

      try {   // Exception Catching Logic
        move = scanner.nextLine().charAt(0);   // Read User Input
      } catch (Exception e) {
        System.out.println("Invalid Move Entry, Try again.");   // Simple Print Error
        continue;  // Go back to "while" on line 33
      }

      int newX = currentX;
      int newY = currentY;

      // Moving the Character
      if (move == 'w') {
        newY--;
      } else if (move == 'a') {
        newX = newX - 3;
      } else if (move == 's') {
        newY++;
      } else if (move == 'd') {
        newX = newX + 3;
      } else {
        System.out.println("Invalid move.");
        continue; // Go back to "while" on line 33
      }

      try {   // Exception Catching Logic
        while (maze[newY][newX] == '#') { // Go to nearest point to the wall
          if (move == 'w') {
            newY++;
          } else if (move == 'a') {
            newX++;
          } else if (move == 's') {
            newY--;
          } else if (move == 'd') {
            newX--;
          } else {
            throw new UnexpectedException("Error with movement");
          }
        }

        // Adjust position to stay within bounds
        newX = Math.max(0, Math.min(newX, maze[0].length - 2));
        newY = Math.max(0, Math.min(newY, maze.length - 2));

        clearConsole();
        maze[currentY][currentX] = ' ';
        currentX = newX;
        currentY = newY;
        maze[currentY][currentX] = 't';
      } catch (ArrayIndexOutOfBoundsException e) {
        newX = Math.max(0, Math.min(newX, maze[0].length - 2));
        newY = Math.max(0, Math.min(newY, maze.length - 2));
      }

      printMaze();
    }

    System.out.println("Congratulations! You solved the maze in " + loopiteration + " loop iterations.");
    scanner.close();
  }

  private static void loadMaze(String filename) {
    try {
      // Get the resource URL directly
      URL resourceUrl = MazeSolver.class.getResource("/mazes/" + filename + ".txt");

      if (resourceUrl == null) {
        System.out.println("Maze file not found: /mazes/" + filename + ".txt");
        System.exit(1);
      }

      // Open the InputStream from the URL
      try (InputStream inputStream = resourceUrl.openStream();
           Scanner fileScanner = new Scanner(inputStream)) {

        int rows = fileScanner.nextInt();
        int cols = fileScanner.nextInt();
        fileScanner.nextLine(); // Consume newline

        maze = new char[rows][cols];
        int lineCount = 0; // Add a line counter for debugging
        for (int i = 0; i < rows; i++) {
          String line = fileScanner.nextLine();
          lineCount++; // Increment the line counter

          // Check if the line is shorter than the expected number of columns
          if (line.length() < cols) {
            System.out.println("Error: Line " + lineCount + " is too short.");
            System.exit(1);
          }

          for (int j = 0; j < cols; j++) {
            maze[i][j] = line.charAt(j);
            if (maze[i][j] == 'S') {
              startX = j;
              startY = i;
            } else if (maze[i][j] == 'E') {
              endX = j;
              endY = i;
            }
          }
        }

        System.out.println("Lines read: " + lineCount); // Print the line count

      } catch (IOException e) {
        System.out.println("Error reading maze file: " + e.getMessage());
        System.exit(1);
      }

    } catch (NullPointerException e) {
      System.out.println("Maze file not found.");
      System.exit(1);
    }
  }

  private static void printMaze() {
    for (int i = 0; i < maze.length; i++) {
      for (int j = 0; j < maze[i].length; j++) {
        System.out.print(maze[i][j]);
      }
      System.out.println();
    }
  }

  // Method to clear the console
  public static void clearConsole() {
    try {
      if (System.getProperty("os.name").contains("Windows")) {
        new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor();

      } else {
        System.out.print("\033[H\033[2J");
        System.out.flush();
      }
    } catch (IOException | InterruptedException e) {
      System.out.println("Error clearing console: " + e.getMessage());
    }
  }
}